flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskManagerMetricGroup.java

```diff
@@ -93,49 +96,24 @@ public TaskMetricGroup addTaskForJob(
 
         String resolvedJobName = jobName == null || jobName.isEmpty() ? jobId.toString() : jobName;
 
-        // we cannot strictly lock both our map modification and the job group modification
-        // because it might lead to a deadlock
-        while (true) {
-            // get or create a jobs metric group
-            TaskManagerJobMetricGroup currentJobGroup;
-            synchronized (this) {
-                currentJobGroup = jobs.get(jobId);
-
-                if (currentJobGroup == null || currentJobGroup.isClosed()) {
-                    currentJobGroup =
-                            new TaskManagerJobMetricGroup(registry, this, jobId, resolvedJobName);
-                    jobs.put(jobId, currentJobGroup);
-                }
-            }
-
-            // try to add another task. this may fail if we found a pre-existing job metrics
-            // group and it is closed concurrently
-            TaskMetricGroup taskGroup =
-                    currentJobGroup.addTask(
-                            jobVertexId, executionAttemptId, taskName, subtaskIndex, attemptNumber);
-
-            if (taskGroup != null) {
-                // successfully added the next task
-                return taskGroup;
-            }
+        TaskManagerJobMetricGroup jobGroup = jobs.get(jobId);
 
-            // else fall through the loop
-        }
-    }
-
-    public void removeJobMetricsGroup(JobID jobId, TaskManagerJobMetricGroup group) {
-        if (jobId == null || group == null || !group.isClosed()) {
-            return;
+        if (jobGroup == null) {
```

Comment:
The check is already done in `close()` method itself (and it cannot be removed from there because of concurrency).
And IIUC, this method `removeJobMetricsGroup` can not be called after `close` because they are called from
`TaskExecutor.releaseJobResources` and `stopTaskExecutorServices` respectively.